packages/@aws-cdk/aws-codeartifact/lib/repository.ts

```diff
@@ -0,0 +1,404 @@
+        principal,
+      ],
+      resources: [resource],
+      actions: iamActions,
+    }));
+
+    this.cfnRepository.permissionsPolicyDocument = p;
+  }
+
+  /**
+     * Adds read actions for the principal to the repository's
+     * resource policy
+     * @param principal The principal for the policy
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/repo-policies.html
+     */
+  public allowReadFromRepository(principal : iam.IPrincipal) : Repository {
+    this.createPolicy(principal,
+      ['codeartifact:DescribePackageVersion',
+        'codeartifact:DescribeRepository',
+        'codeartifact:GetPackageVersionReadme',
+        'codeartifact:GetRepositoryEndpoint',
+        'codeartifact:ListPackageVersionAssets',
+        'codeartifact:ListPackageVersionDependencies',
+        'codeartifact:ListPackageVersions',
+        'codeartifact:ListPackages',
+        'codeartifact:ReadFromRepository'],
+    );
+    return this;
+  }
+
+  /**
+     * Allows PublishPackageVersion and PutPackageMetadata only for the package and namespace
+     * when acted upon by the principal.
+     * @param principal The principal for the policy
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/repo-policies.html
+     */
+  public allowWriteToRepositoryPackage(principal : iam.IPrincipal, repositoryPackage: PolicyRepositoryPackage) : Repository {
+    this.allowWriteToRepository(principal,
+      `arn:aws:codeartifact:${Aws.REGION}:${this.repositoryDomainOwner}:package/${this.repositoryDomainName}/${this.repositoryName}/${repositoryPackage.packageFormat}/${repositoryPackage.packageNamespace}/${repositoryPackage.packageName}`,
+    );
+    return this;
+  }
+
+  /**
+     * Adds PublishPackageVersion and PutPackageMetadata for the principal to the repository's
+     * resource policy
+     * @param principal The principal for the policy
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/repo-policies.html
+     */
+  public allowWriteToRepository(principal : iam.IPrincipal, resource : string = '*') : Repository {
+    this.createPolicy(principal,
+      ['codeartifact:PublishPackageVersion',
+        'codeartifact:PutPackageMetadata'],
+      resource,
+    );
+    return this;
+  }
+  /**
+     * Adds DeletePacakgeVersion for the principal to the repository's
+     * resource policy
+     * @param principal The principal for the policy
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/repo-policies.html
+     */
+  public allowDeleteFromRepository(principal : iam.IPrincipal) : Repository {
+    this.createPolicy(principal,
+      ['codeartifact:DeletePackageVersion'],
+    );
+    return this;
+  }
+
+  /**
+     * External connections to pull from
+     * @default None
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/external-connection.html#adding-an-external-connection
+   */
+  public withExternalConnections(...externalConnections: e.ExternalConnection[]): IRepository {
```

Comment:
Is there a compelling reason to include these post-construction -- for both this and `withUpstream`? I don't see the value off the top.

Reply:
These could be added at later part of the flow based on some additional logic. It's assumptive at best. It also tries to stay in the functional pattern. We could remove it for now.